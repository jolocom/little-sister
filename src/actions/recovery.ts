import { ThunkAction } from '../store'
//import { SoftwareKeyProvider } from 'jolocom-lib/js/vaultedKeyProvider/softwareProvider'
import { navigationActions, genericActions } from './index'
import { routeList } from '../routeList'
import settingKeys from '../ui/settings/settingKeys'
import { removeNotification, scheduleNotification } from './notifications'
import { entropyToMnemonic, mnemonicToEntropy } from 'bip39'
 // TODO Import ^ from jolocom-lib
import useResetKeychainValues from 'src/ui/deviceauth/hooks/useResetKeychainValues'
import { PIN_SERVICE } from 'src/ui/deviceauth/utils/keychainConsts'
import { createWarningNotification } from 'src/lib/notifications'
import strings from 'src/locales/strings'
import I18n from 'src/locales/i18n'

export const showSeedPhrase = (): ThunkAction => async (
  dispatch,
  getState,
  agent,
) => {
  const encryptedSeed = await agent.storage
    .get.setting('encryptedSeed')

  if (!encryptedSeed) {
    throw new Error('Can not retrieve Seed from database')
  }

  const decrypted = await agent.idw.asymDecrypt(
    Buffer.from(encryptedSeed.b64Encoded, 'base64'),
    await agent.passwordStore.getPassword()
  )

  return dispatch(
    navigationActions.navigate({
      routeName: routeList.SeedPhrase,
      params: { mnemonic: entropyToMnemonic(decrypted)},
    }),
  )
}

/**
 * called during PIN code restoration
 */
export const onRestoreAccess = (mnemonicInput: string[]): ThunkAction => async (
  dispatch,
  getState,
  agent,
) => {
  let recovered = false

  try {
    const recoveredEntropy = Buffer.from(
      mnemonicToEntropy(mnemonicInput.join(' ')),
      'hex'
    )
    if (agent.didMethod.recoverFromSeed) {
      const { identityWallet } = await agent.didMethod.recoverFromSeed(
        recoveredEntropy,
        await agent.passwordStore.getPassword()
      )
      recovered = identityWallet.did === agent.idw.did
    }
  } catch(e) {
    console.error('onRestoreAccess failed', e)
  }

  if (recovered) {
    const resetServiceValuesInKeychain = useResetKeychainValues(PIN_SERVICE)
    await resetServiceValuesInKeychain()
    dispatch(navigationActions.navigatorResetHome())

    // we re-lock the app, which will trigger the create pin screen
    return dispatch(genericActions.lockApp())
  } else {
    const notification = createWarningNotification({
      title: I18n.t(strings.AWKWARD),
      message: I18n.t(strings.IT_SEEMS_LIKE_WE_CANT_DO_THIS)
    })
    return dispatch(scheduleNotification(notification))
  }
}

export const setSeedPhraseSaved = (): ThunkAction => async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  // No delete call available yet, overwriting with empty object
  await backendMiddleware.storage.store.setting('encryptedSeed', {})
  await backendMiddleware.storage.store.setting(
    settingKeys.seedPhraseSaved,
    true,
  )

  const stickies = getState().notifications.queue.filter(n => !n.dismiss)

  stickies.map(sticky => dispatch(removeNotification(sticky)))

  return dispatch({
    type: 'SET_SEED_PHRASE_SAVED',
  })
}
