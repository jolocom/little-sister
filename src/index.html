<!DOCTYPE html>

<html>
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport"
          content="width=device-width, initial-scale=1, user-scalable=no">

    <title>Little Sister</title>

    <link href="https://fonts.googleapis.com/css?family=Roboto:100,400,300,600"
          rel="stylesheet"
          type="text/css" />
    <link href="https://fonts.googleapis.com/icon?family=Material+Icons"
          rel="stylesheet">

    <style type="text/css">
      * {
        margin: 0;
        padding: 0;
      }
      html, body {
        height: 100%;
      }
      body {
        font-family: Roboto;
      }
      #app {
        width: 100%;
        height: 100%;
      }
    </style>

    <!-- <meta http-equiv="Content-Security-Policy" content="default-src * data: gap: https://ssl.gstatic.com 'unsafe-eval'; style-src * 'unsafe-inline'; media-src *; img-src * data:"> -->
    <meta name="format-detection" content="telephone=no">
    <meta name="msapplication-tap-highlight" content="no">
    <meta name="viewport" content="user-scalable=no, initial-scale=1, maximum-scale=1, minimum-scale=1, width=device-width">
  </head>
  <body>
    <div id="app"></div>
    <script src="https://solid.github.io/releases/rdflib.js/rdflib-0.7.0.min.js"></script>
    <script src="https://d3js.org/d3.v4.min.js"></script>
    <script src="js/bundle.js"></script>
  </body>
</html>
