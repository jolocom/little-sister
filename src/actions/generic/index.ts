import { AppError, ErrorCode } from '../../lib/errors'
import I18n from '../../locales/i18n'
import { routeList } from '../../routeList'
import { navigationActions } from '../../actions'
import { ThunkAction } from '../../store'
import settingKeys from '../../ui/settings/settingKeys'

import * as Keychain from 'react-native-keychain'
// TODO use the settings items from storage
import AsyncStorage from '@react-native-community/async-storage'
// TODO don't depend on the crypto lib, perhaps use the rust crypto utils?
import crypto from 'crypto'

import { termsOfServiceDE } from 'src/ui/termsAndPrivacy/legalTexts'

import {
  AppWrapConfig,
  APPWRAP_UPDATE_CONFIG,
  APPWRAP_SHOW_LOADER,
  APPWRAP_REGISTER_CONFIG,
  APPWRAP_UNREGISTER_CONFIG,
  APPWRAP_SET_LOCKED,
  APPWRAP_SET_DISABLE_LOCK,
} from '../../reducers/generic'
import { AnyAction } from 'redux'
import { PIN_SERVICE } from 'src/ui/deviceauth/utils/keychainConsts'

// Default delay on the loading state value before it can switch back to 'false'
const DEFAULT_LOADING_LATCH_DELAY_MS = 300

export const showErrorScreen = (
  error: AppError | Error,
): ThunkAction => dispatch => {
  const appError: AppError =
    error.constructor === AppError
      ? (error as AppError)
      : new AppError(ErrorCode.Unknown, error)

  return dispatch(
    navigationActions.navigate({
      routeName: routeList.Exception,
      params: {
        returnTo: appError.navigateTo,
        error: appError,
      },
    }),
  )
}

const hashString = (text: string) => {
  return crypto
    .createHash('sha256')
    .update(text)
    .digest('hex')
}

const TERMS_OF_CONDITIONS_KEY = 'TERMS_OF_CONDITIONS'
export const checkTermsOfService = (
  route: routeList,
  onSubmit?: () => void,
): ThunkAction => async dispatch => {
  const storageHash = await AsyncStorage.getItem(TERMS_OF_CONDITIONS_KEY)
  const currentHash = hashString(termsOfServiceDE)
  const shouldShowTerms = storageHash !== currentHash

  if (!shouldShowTerms && onSubmit) onSubmit()

  if (shouldShowTerms) {
    return dispatch(
      navigationActions.navigate({
        routeName: routeList.TermsOfServiceConsent,
        params: { nextRoute: route, onSubmit },
      })
    )
  } else {
    return dispatch(
      navigationActions.navigate({
        routeName: route,
      })
    )
  }
}

export const storeTermsOfService = (
  route: routeList,
): ThunkAction => async dispatch => {
  const termsHash = hashString(termsOfServiceDE)
  await AsyncStorage.setItem(TERMS_OF_CONDITIONS_KEY, termsHash)

  dispatch(navigationActions.navigate({ routeName: route }))
}

export const setLocale = (locale: string): ThunkAction => async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  await backendMiddleware.storage.store.setting(settingKeys.locale, {
    selected: locale
  })

  I18n.locale = locale
  dispatch({ type: 'SET_LOCALE', value: locale })

  // we need to reset the navigator so that all screens are re-rendered with the
  // new locale
  return dispatch(navigationActions.navigatorReset())
}

const setLoading = (value: boolean) => ({
  type: APPWRAP_SHOW_LOADER,
  value,
})

export const setLocked = (value: boolean) => ({
  type: APPWRAP_SET_LOCKED,
  value,
})

export const setDisableLock = (value: boolean) => ({
  type: APPWRAP_SET_DISABLE_LOCK,
  value,
})

export const lockApp = (
): ThunkAction => async (dispatch, getState) => {
  // if the lock is globally disabled, or there is no identity created/loaded
  // yet, then this action has no effect
  const disableLock = getState().generic.disableLock
  const hasIdentity = !!getState().account.did.did
  if (disableLock || !hasIdentity) return

  // otherwise we lock the app, using the appropriate lock screen
  const keychainPin = await Keychain.getGenericPassword({
    service: PIN_SERVICE,
  })
  const lockScreen = keychainPin ? routeList.Lock : routeList.RegisterPIN

  dispatch(setLocked(true))
  return dispatch(navigationActions.navigate({ routeName: lockScreen }))
}

export const unlockApp = (
  pin: string,
  resetNav?: boolean,
): ThunkAction => async dispatch => {
  const keychainPin = await Keychain.getGenericPassword({
    service: PIN_SERVICE,
  })

  if (!keychainPin || keychainPin.password !== pin) throw new Error('wrongPin')

  dispatch(setLocked(false))
  return resetNav
    ? dispatch(navigationActions.navigatorResetHome())
    : dispatch(navigationActions.navigateBack())
}

export const updateAppWrapConfig = (value: AppWrapConfig) => ({
  type: APPWRAP_UPDATE_CONFIG,
  value,
})
export const registerAppWrapConfig = (value: AppWrapConfig & {}) => ({
  type: APPWRAP_REGISTER_CONFIG,
  value,
})
export const unregisterAppWrapConfig = (value: AppWrapConfig) => ({
  type: APPWRAP_UNREGISTER_CONFIG,
  value,
})

export const showAppLoading = (
  shouldShow: boolean,
  latchDelay = DEFAULT_LOADING_LATCH_DELAY_MS,
): ThunkAction => async (dispatch, getState) => {
  const wasShowing = getState().generic.appWrapConfig.loading
  if (shouldShow && !wasShowing) {
    // for setting to "true" we do it immediately and set current state
    const ret = dispatch(setLoading(shouldShow))
    return new Promise<AnyAction>(resolve => {
      setTimeout(() => resolve(ret), latchDelay)
    })
  }

  if (wasShowing && !shouldShow) {
    // for setting to 'false', we delay by the latchDelay amount
    return dispatch(delayedSetLoading(shouldShow, latchDelay))
  }
}

let nextLoadingState = false
let timeout: number | undefined
const delayedSetLoading = (
  value: boolean,
  latchDelay: number,
): ThunkAction => async dispatch => {
  nextLoadingState = value
  if (timeout !== undefined) clearTimeout(timeout)
  timeout = setTimeout(() => {
    dispatch(setLoading(nextLoadingState))
    timeout = undefined
  }, latchDelay)
}
