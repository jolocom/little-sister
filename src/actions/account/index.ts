import { groupBy, map, mergeRight, omit, uniq, zipWith } from 'ramda'
import { SignedCredential } from 'jolocom-lib/js/credentials/signedCredential/signedCredential'
import { IdentitySummary, CredentialMetadataSummary } from '@jolocom/sdk'

import { navigationActions } from 'src/actions/'

import { routeList } from 'src/routeList'
import { CategorizedClaims, DecoratedClaims } from 'src/reducers/account'
import {
  getClaimMetadataByCredentialType,
  getCredentialUiCategory,
  getUiCredentialTypeByType,
} from 'src/lib/util'
import { ThunkAction } from 'src/store'
import { compose } from 'redux'
import { Not } from 'typeorm'
import { HAS_EXTERNAL_CREDENTIALS } from './actionTypes'
import { SDKError } from '@jolocom/sdk'
import { checkTermsOfService } from '../generic'
import { checkRecoverySetup } from '../notifications/checkRecoverySetup'

export const setDid = (did: string) => ({
  type: 'DID_SET',
  value: did,
})

export const setSelected = (claim: DecoratedClaims) => ({
  type: 'SET_SELECTED',
  selected: claim,
})

export const resetSelected = () => ({
  type: 'RESET_SELECTED',
})

export const handleClaimInput = (fieldValue: string, fieldName: string) => ({
  type: 'HANDLE_CLAIM_INPUT',
  fieldName,
  fieldValue,
})

export const checkIdentityExists: ThunkAction = async (
  dispatch,
  getState,
  agent,
) => {
  try {
    const identityWallet = await agent.loadIdentity()
    const userDid = identityWallet.identity.did
    dispatch(setDid(userDid))
    await dispatch(setClaimsForDid)
    await dispatch(checkRecoverySetup)
    return dispatch(checkTermsOfService(routeList.Home))
  } catch (err) {
    if (
      err.message === SDKError.codes.NoEntropy ||
      err.message === SDKError.codes.NoWallet
    ) {
      // No seed in database, user must register
      // But check if a registration was already in progress
      const isRegistering = getState().registration.loading.isRegistering

      const routeName = isRegistering
        ? routeList.RegistrationProgress
        : routeList.Landing

      return dispatch(navigationActions.navigate({ routeName }))
    }

    throw err
  }
}

export const openClaimDetails = (
  claim: DecoratedClaims,
): ThunkAction => dispatch => {
  dispatch(setSelected(claim))
  return dispatch(
    navigationActions.navigate({
      routeName: routeList.ClaimDetails,
    }),
  )
}

export const saveClaim: ThunkAction = async (
  dispatch,
  getState,
  agent,
) => {
  const { identityWallet, storage, passwordStore } = agent

  const did = getState().account.did.did
  const claimsItem = getState().account.claims.selected
  const password = await passwordStore.getPassword()

  const verifiableCredential = await identityWallet.create.signedCredential(
    {
      metadata: getClaimMetadataByCredentialType(claimsItem.credentialType),
      // the library acts directly on the object passed in, so a copy should be made first
      claim: { ...claimsItem.claimData },
      subject: did,
    },
    password,
  )

  if (claimsItem.id) {
    await storage.delete.verifiableCredential(claimsItem.id)
  }

  await storage.store.verifiableCredential(verifiableCredential)

  await dispatch(setClaimsForDid)

  return dispatch(navigationActions.navigatorResetHome())
}

export const hasExternalCredentials: ThunkAction = async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  const { storage, identityWallet } = backendMiddleware
  // TODO FIXME
  // we only need a count, no need to actually load and deserialize
  // all of them
  const externalCredentials = await storage.get.verifiableCredential({
    issuer: Not(identityWallet.did),
  })

  return dispatch({
    type: HAS_EXTERNAL_CREDENTIALS,
    value: externalCredentials.length !== 0,
  })
}
export const setClaimsForDid: ThunkAction = async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  const { storage } = backendMiddleware

  const verifiableCredentials: SignedCredential[] = await storage.get.verifiableCredential()

  const metadata = await Promise.all(
    verifiableCredentials.map(el => storage.get.credentialMetadata(el)),
  )

  const issuers = uniq(verifiableCredentials.map(cred => cred.issuer))

  const issuerMetadata = await Promise.all(
    issuers.map(storage.get.publicProfile),
  )

  const claims = prepareClaimsForState(
    verifiableCredentials,
    metadata,
    issuerMetadata,
  ) as CategorizedClaims

  return dispatch({
    type: 'SET_CLAIMS_FOR_DID',
    claims,
  })
}

export const prepareClaimsForState = (
  credentials: SignedCredential[],
  credentialMetadata: Array<CredentialMetadataSummary | {}>,
  issuerMetadata: IdentitySummary[],
) =>
  compose(
    groupBy(getCredentialUiCategory),
    zipWith(mergeRight, credentialMetadata),
    map(addIssuerInfo(issuerMetadata)),
    map(convertToDecoratedClaim),
  )(credentials)

export const addIssuerInfo = (issuerProfiles: IdentitySummary[]) => (
  claim: DecoratedClaims,
) => {
  if (!issuerProfiles || !issuerProfiles.length) {
    return claim
  }

  const issuer = issuerProfiles.find(el => el.did === claim.issuer.did)

  return issuer
    ? {
        ...claim,
        issuer,
      }
    : claim
}

/** @TODO Util, make subject mandatory (in lib) */
export const convertToDecoratedClaim = ({
  claim,
  type,
  issuer: issuerDid,
  id,
  expires,
}: SignedCredential): DecoratedClaims => ({
  credentialType: getUiCredentialTypeByType(type),
  issuer: {
    did: issuerDid,
  },
  claimData: omit(['id'], claim),
  id,
  subject: claim.id || 'Not found',
  expires: expires || undefined,
})
