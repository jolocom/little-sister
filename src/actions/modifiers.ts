import { ActionCreator } from 'redux'
import NetInfo from '@react-native-community/netinfo'
import { ThunkAction } from 'src/store'
import { AppError } from '../lib/errors'
import { showErrorScreen, showAppLoading } from './generic'
import { scheduleNotification } from './notifications'
import { createInfoNotification } from '../lib/notifications'
import I18n from 'src/locales/i18n'
import strings from '../locales/strings'
import { scheduleOfflineNotification } from '.'

/**
 * Curried function that wraps a {@link ThunkAction} with two calls to the provided loadingAction
 * {@link ActionCreator} (called with "true" before the wrapped action is dispatched, and with "false" after)
 * @param loadingAction - An action creator to create the actions to be dispatched before and after the wrappedAction-
 * @param wrappedAction - The thunkAction to be wrapped
 * @example dispatch(withLoading(showAppLoading)(saveClaims))
 */
export const withLoadingHandler = (
  loadingAction: ActionCreator<ThunkAction>,
) => (wrappedAction: ThunkAction): ThunkAction => async dispatch => {
  try {
    dispatch(loadingAction(true))
    return await dispatch(wrappedAction)
  } finally {
    dispatch(loadingAction(false))
  }
}

type ErrorModifier = (error: AppError | Error) => AppError

/**
 * Curried function that wraps a {@link ThunkAction} with the provided error handler to be dispatched on thrown error
 * @param errorHandler - An {@link ActionCreator} to be dispatched with the error if an error is thrown
 * @param errorModifier - An optional modifier for the error which is thrown, is Identity function by default
 * @param wrappedAction - The {@link ThunkAction} to be wrapped. If dispatching it throws an error, the errorhandler will be dispatched
 * @example dispatch(withErrorHandling(showErrorScreen)(saveClaims))
 * @dev The return value from this modifier can be passed to other modifiers, i.e. modifiers can be composed
 * e.g. dispatch(withLoading(showAppLoading)(withErrorHandling(showErrorScreen)(saveClaims)))
 */
export const withErrorHandler = (
  errorHandler: ActionCreator<ThunkAction>,
  errorModifier: ErrorModifier | undefined = undefined,
) => (
  wrappedAction: ThunkAction,
  modifier: ErrorModifier | undefined = undefined,
): ThunkAction => async dispatch => {
  modifier = modifier || errorModifier
  try {
    return await dispatch(wrappedAction)
  } catch (error) {
    if (modifier) error = modifier(error)
    return dispatch(errorHandler(error))
  }
}

/**
 * Curried function that wraps a {@link ThunkAction} with a notification on internet connection absence
 * @param wrappedAction - The thunkAction to be wrapped
 * @example dispatch(withInternet((saveClaims))
 */
export const withInternet = (
  wrappedAction: ThunkAction,
): ThunkAction => async dispatch => {
  const state = await NetInfo.fetch()
  if (!state.isConnected) {
    return dispatch(
      scheduleOfflineNotification(
        strings.PLEASE_CHECK_YOUR_CONNECTION_AND_TRY_AGAIN,
      ),
    )
  }
  return dispatch(wrappedAction)
}

export const withLoading = withLoadingHandler(showAppLoading)
export const withErrorScreen = withErrorHandler(showErrorScreen)
