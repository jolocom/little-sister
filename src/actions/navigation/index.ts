import {
  NavigationAction,
  NavigationActions,
  NavigationContainerComponent,
  NavigationNavigateAction,
  NavigationNavigateActionPayload,
  NavigationResetActionPayload,
  NavigationRouter,
  StackActions,
} from 'react-navigation'
import { routeList } from 'src/routeList'
import { AppError, ErrorCode } from '../../lib/errors'
import { ThunkAction } from 'src/store'
import { consumeInteractionToken } from '../sso/consumeInteractionToken'

const deferredNavActions: NavigationAction[] = []
let dispatchNavigationAction = (action: NavigationAction) => {
    deferredNavActions.push(action)
  },
  navigator: NavigationContainerComponent

export const setTopLevelNavigator = (nav: NavigationContainerComponent) => {
  dispatchNavigationAction = nav.dispatch.bind(nav)
  navigator = nav
  deferredNavActions.forEach(dispatchNavigationAction)
  deferredNavActions.length = 0
}

/**
 * NOTE: navigate and navigatorReset both dispatch the navigation actions but
 * the actions are not handled by our reducers. Dispatching is useful for testing
 * (comparing snapshots of store actions) and it makes typescript happy
 */
export const navigate = (
  options: NavigationNavigateActionPayload,
  replace = false
): ThunkAction => dispatch => {
  const action = replace
    ? StackActions.replace(options)
    : NavigationActions.navigate(options)

  dispatchNavigationAction(action)
  return dispatch(action)
}

export const navigateBack = (): ThunkAction => dispatch => {
  const action = NavigationActions.back()
  dispatchNavigationAction(action)
  return dispatch(action)
}

export const navigateBackHome = (): ThunkAction => dispatch => {
  dispatch(navigateBack())
  return dispatch(navigatorResetHome())
}

export const navigatorReset = (
  newScreen?: NavigationNavigateActionPayload,
): ThunkAction => dispatch => {
  const resetActionPayload: NavigationResetActionPayload = {
    index: 0,
    actions: [],
  }

  if (newScreen) {
    resetActionPayload.actions.push(NavigationActions.navigate(newScreen))
  } else {
    // @ts-ignore
    const navState = navigator.state.nav
    // @ts-ignore
    const navRouter: NavigationRouter = navigator._navigation.router

    if (navRouter) {
      const { path, params } = navRouter.getPathAndParamsForState(navState)
      const action = navRouter.getActionForPathAndParams(
        path,
        params,
      ) as NavigationNavigateAction

      // getActionForPathAndParams is typed to potentially return null, but we are
      // using it on the current state itself, so this should "never" happen
      if (!action) {
        throw new Error('impossible')
      }

      if (action.action) {
        // since the top level router is a SwitchRouter,
        // the first action will be to navigate to MainStack, but we are using a
        // StackReset action which will be caught by MainStack, so we can just pass
        // in the nested Navigate action
        resetActionPayload.actions.push(action.action)
      }
    }
  }

  const resetAction = StackActions.reset(resetActionPayload)
  dispatchNavigationAction(resetAction)
  return dispatch(resetAction)
}

export const navigatorResetHome = (): ThunkAction => dispatch =>
  dispatch(navigatorReset({ routeName: routeList.Home }))

/**
 * The function that parses a deep link to get the route name and params
 * It then matches the route name and dispatches a corresponding action
 * @param url - a deep link string with the following schema: appName://routeName/params
 */
export const handleDeepLink = (url: string): ThunkAction => (
  dispatch,
  getState,
  agent,
) => {
  // The identityWallet is initialised before the deep link is handled. If it
  // is not initialized, then we may not even have an identity.
  if (!agent.identityWallet) {
    return dispatch(
      navigate({
        routeName: routeList.Landing,
      }),
    )
  }

  let routeName = '',
    params = ''

  if (url) {
    const [scheme, uri] = url.split('://')
    if (scheme && uri) {
      const parts = uri.split('/')
      if (scheme.startsWith('http')) {
        routeName = parts[1]
        params = parts[2]
      } else {
        routeName = parts[0]
        params = parts[1]
      }
    }
  }

  const supportedRoutes = ['consent', 'authenticate']
  if (supportedRoutes.includes(routeName)) {
    return dispatch(consumeInteractionToken(params))
  } else {
    throw new AppError(
      ErrorCode.DeepLinkUrlNotFound,
      new Error('Could not handle interaction token'),
    )
  }
}
