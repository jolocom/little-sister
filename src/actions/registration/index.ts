import * as loading from 'src/actions/registration/loadingStages'
import { setDid } from 'src/actions/account'
import { navigatorResetHome } from '../navigation'
import { setSeedPhraseSaved } from '../recovery'
import { generateSecureRandomBytes } from '@jolocom/sdk/js/util'
import { ThunkAction } from '../../store'
import { entropyToMnemonic } from 'bip39'
import { genericActions, navigationActions } from '..'
import useResetKeychainValues from 'src/ui/deviceauth/hooks/useResetKeychainValues'
import { PIN_SERVICE } from 'src/ui/deviceauth/utils/keychainConsts'
import { createWarningNotification } from 'src/lib/notifications'
import { scheduleNotification } from '../notifications'
import strings from 'src/locales/strings'
import I18n from 'i18n-js'

const humanTimeout = () => new Promise(resolve => setTimeout(resolve, 1000))

export const setLoadingMsg = (loadingMsg: string) => ({
  type: 'SET_LOADING_MSG',
  value: loadingMsg,
})

export const setIsRegistering = (value: boolean) => ({
  type: 'SET_IS_REGISTERING',
  value,
})

export const createIdentity = (encodedEntropy: string): ThunkAction => async (
  dispatch,
  getState,
  agent,
) => {
  const isRegistering = getState().registration.loading.isRegistering
  if (isRegistering) return

  dispatch(setIsRegistering(true))

  // strings.REGISTERING_DECENTRALIZED_IDENTITY
  // aka "we are generating a random number"
  dispatch(setLoadingMsg(loading.loadingStages[0]))

  const seed = await generateSecureRandomBytes(16)
  const identity = await agent.loadFromMnemonic(entropyToMnemonic(seed))
  // and it's too fast so slow down
  await humanTimeout()

  // strings.ENCRYPTING_AND_STORING_DATA_LOCALLY
  dispatch(setLoadingMsg(loading.loadingStages[1]))

  // TODO Better call here.
  const encryptedSeed = await identity.asymEncryptToDid(
    Buffer.from(seed),
    identity.did, {
      prefix: '',
      resolve: async _ => identity.identity
    })

  await agent.storage.store.setting(
    'encryptedSeed',
    {
      b64Encoded: encryptedSeed.toString('base64')
    }
  )
  await humanTimeout()

  // strings.PREPARING_LAUNCH
  dispatch(setLoadingMsg(loading.loadingStages[2]))
  dispatch(setDid(identity.did))
  await humanTimeout()
  dispatch(setIsRegistering(false))

  // clear the saved PIN code, if any
  await useResetKeychainValues(PIN_SERVICE)()

  return dispatch(genericActions.lockApp())
}

export const recoverIdentity = (mnemonic: string): ThunkAction => async (
  dispatch,
  getState,
  agent,
) => {
  dispatch(setIsRegistering(true))
  try {
    const identity = await agent.loadFromMnemonic(mnemonic);

    dispatch(setDid(identity.did))
    await dispatch(setSeedPhraseSaved())

    await dispatch(navigatorResetHome())
    await dispatch(genericActions.lockApp())

    return dispatch(setIsRegistering(false))
  } catch (e) {
    const notification = createWarningNotification({
      title: I18n.t(strings.AWKWARD),
      message: I18n.t(strings.IT_SEEMS_LIKE_WE_CANT_DO_THIS)
    })
    dispatch(scheduleNotification(notification))
    dispatch(navigationActions.navigateBack())
    return dispatch(setIsRegistering(false))
  }
}
