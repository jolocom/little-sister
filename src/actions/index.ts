import { Interaction, FlowType } from '@jolocom/sdk'
import { ThunkAction } from 'src/store'

import * as accountActions from 'src/actions/account/'
import * as registrationActions from 'src/actions/registration/'
import * as navigationActions from 'src/actions/navigation/'
import * as genericActions from 'src/actions/generic/'
import * as ssoActions from 'src/actions/sso/'
import * as recoveryActions from 'src/actions/recovery'
import * as notificationsActions from 'src/actions/notifications'
import strings from 'src/locales/strings'
import I18n from 'src/locales/i18n'
import { createInfoNotification } from '../lib/notifications'

export {
  accountActions,
  registrationActions,
  navigationActions,
  genericActions,
  ssoActions,
  recoveryActions,
  notificationsActions,
}

export const scheduleOfflineNotification = (
  message: string,
): ThunkAction => dispatch =>
  dispatch(
    notificationsActions.scheduleNotification(
      createInfoNotification({
        title: I18n.t(strings.NO_INTERNET_CONNECTION),
        message: I18n.t(message),
      }),
    ),
  )

export const interactionHandlers: {
  [flowType: string]: (interxn: Interaction) => ThunkAction
} = {
  [FlowType.Authentication]: ssoActions.consumeAuthenticationRequest,
  [FlowType.CredentialOffer]: ssoActions.consumeCredentialOfferRequest,
  [FlowType.CredentialShare]: ssoActions.consumeCredentialRequest,
  [FlowType.EstablishChannel]: ssoActions.consumeEstablishChannelRequest,
  [FlowType.Resolution]: ssoActions.consumeResolutionRequest,
}
