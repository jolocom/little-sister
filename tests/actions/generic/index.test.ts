import { genericActions } from 'src/actions'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'

describe('Generic action creators', () => {
  describe('showErrorScreen', () => {
    it('should navigate to error screen and provide the message', () => {
      const mockStore = configureStore([thunk])({})
      const mockError = {
        message: 'MOCK BAD ERROR',
        stack: 'MOCK STACK TRACE'
      }

      const action = genericActions.showErrorScreen(mockError)

      action(mockStore.dispatch)
      expect(mockStore.getActions()).toMatchSnapshot()
    })
  })
})
