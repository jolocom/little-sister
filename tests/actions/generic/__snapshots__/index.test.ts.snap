// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generic action creators showErrorScreen should navigate to error screen and provide the message 1`] = `
Array [
  Object {
    "params": Object {
      "error": [Error: Unknown Error],
      "returnTo": "Home",
    },
    "routeName": "Exception",
    "type": "Navigation/NAVIGATE",
  },
]
`;
